File: UserHashTable.py
class UserHashTable:
    def __init__(self, size=100):
        # Initialize the hash table with empty lists for each bucket
        self.size = size
        self.table = [[] for _ in range(size)]

    def _hash(self, username):
        # Compute the hash index based on the username
        return hash(username) % self.size

    def add_user(self, username: str, profile_data: dict):
        # Find the index for the username
        index = self._hash(username)
        # Check if username already exists in the chain
        for i, (user, data) in enumerate(self.table[index]):
            if user == username:
                # Update profile data if username already exists
                self.table[index][i] = (username, profile_data)
                return
        # If not found, append new username and profile_data
        self.table[index].append((username, profile_data))

    def get_user(self, username: str) -> dict:
        # Find the index for the username
        index = self._hash(username)
        # Search for the username in the chain at that index
        for user, data in self.table[index]:
            if user == username:
                return data  # Return profile data if found
        return None  # Return None if user not found

File: UserHashTable.cpp
#include <vector>
#include <list>
#include <string>
#include <functional>  // for std::hash
#include <utility>     // for std::pair

class UserHashTable {
public:
    UserHashTable(size_t size = 100) : table(size) {}

    void addUser(const std::string& username, const std::string& profileData) {
        size_t index = hashFunction(username);
        // Check if the username already exists in the chain
        for (auto& pair : table[index]) {
            if (pair.first == username) {
                pair.second = profileData;  // Update profile data if username exists
                return;
            }
        }
        // If not found, add new username and profileData pair
        table[index].emplace_back(username, profileData);
    }

    std::string getUser(const std::string& username) const {
        size_t index = hashFunction(username);
        // Search for the username in the chain at that index
        for (const auto& pair : table[index]) {
            if (pair.first == username) {
                return pair.second;  // Return profile data if found
            }
        }
        return "";  // Return empty string if user not found
    }

private:
    size_t hashFunction(const std::string& username) const {
        return std::hash<std::string>{}(username) % table.size();
    }

    std::vector<std::list<std::pair<std::string, std::string>>> table;  // Hash table with separate chaining
};
